### PR TITLE
CBL-5514: Swift MutableDocument's collection is not set when a new document is saved

### PR DESCRIPTION
**Pull request by velicuvlad**
- change Document.collection API to have a private setter
- set collection on save
- add 1 tests for each side (Obj-C and Swift)

### CODE REVIEW

**Comment by velicuvlad** [comment]
Tests failing is unrelated to this PR. Somewhat flaky as it seems the timeout period is not big enough... all of them pass on local. Will investigate as part of the code coverage ticket.

**Review by bmeike — APPROVED** [review]
Be sure you get confirmation from someone who actually knows what they are doing, for this.  There was *lots* of discussion during the original implementation of Collections, about whether the platform code was allowed to hold the reference to one, or not.  I believe the issue is that the thing to which you hold a reference might become invalid, behind your back.    Worth checking.

**Comment by velicuvlad** [comment]
I can confirm that this is behaviour is already there for Obj-C. The Obj-C tests are just to confirm the behaviour.

A new MutableDoc is getting created without any collection ref, meanwhile it gets one when it is saved (as part of the process). I initially thought the change would be bigger as I didn't know that Swift has private setters (Pasin made me aware - kudos). For Swift the collection attribute was non-modifiable as it was a `let`.

What exactly do you refer as getting invalid? The C4Collection object? We do not allow the change of `doc.collection` so in the case you want to move a document, a brand new copy must be made anyway... I would expect that if the collection is invalid by magic, as it cannot be changed, that document is pretty much orphan and the solution becomes the same -> make a brand new copy of it

When we pull a doc out of the db, it does come attached with a collection so we already hold ref to it. I can't think of a case for which that collection would be invalided at the exact time of the save, when save is a collection operation - so if doesn't exist it will fail on save in the first place

File: Swift/Collection.swift [hunk 1]
```diff
@@ -107,6 +107,9 @@ public final class Collection : CollectionChangeObservable, Indexable, Equatable
     /// Throws an NSError with the CBLError.notOpen code, if the collection is deleted or
     /// the database is closed.
     public func save(document: MutableDocument) throws {
+        if (document.collection == nil){
+            document.collection = self
+        }
```
Comment (pasin) [inline Swift/Collection.swift:112]: This should be done after saving is successful.

File: Swift/Collection.swift [hunk 2]
```diff
@@ -107,6 +107,9 @@ public final class Collection : CollectionChangeObservable, Indexable, Equatable
     /// Throws an NSError with the CBLError.notOpen code, if the collection is deleted or
     /// the database is closed.
     public func save(document: MutableDocument) throws {
+        if (document.collection == nil){
```
Comment (pasin) [inline Swift/Collection.swift:110]: Format: A space after ).

File: Swift/Collection.swift (hunk 1 above)
Comment (pasin) [reply Swift/Collection.swift:112]: There are another two save functions that need set the collection to the saved doc as well.

**Review by pasin — CHANGES_REQUESTED** [review]

**Review by pasin — APPROVED** [review]

**Pull request merged**